### ISSUE TITLE
BSP sets the source folder wrong in the context of the sources' package in IntelliJ

### ISSUE THREAD

**Issue by Gedochao** [VirtusLab/scala-cli#1324]
**Version(s)**
0.1.12

**Describe the bug**
Scala CLI's BSP marks the sources directory in IDEA IntelliJ to the parent directory of the sources, regardless of the package structure. This messes up the IntelliJ project structure and makes navigating the sources less intuitive. Technically, everything will compile and run as normal, but it's not great UX when working in IntelliJ.

**To Reproduce**
For a structure like this
```
▶ tree -A
.
└── src
    └── Main.scala

1 directory, 1 file
```
I setup `BSP` with `scala-cli`
```
scala-cli setup-ide .
```
then, I add a package declaration in `Main.scala` and decide to mirror the package structure in directories.
<img width="1164" alt="image" src="https://user-images.githubusercontent.com/18601388/188572312-534b9f8a-3007-4623-93d4-54f007fbfadd.png">
As a result, I have a structure like this:
```
▶ tree -A              
.
└── src
    └── com
        └── github
            └── gedochao
                └── Main.scala
```
<img width="1013" alt="image" src="https://user-images.githubusercontent.com/18601388/188572590-f61a5f60-dd0b-496e-b24e-114041b0c203.png">

Now, a workspace reload with Scala CLI's BSP messes up my project structure by changing the sources directory to the immediate parent directory of `Main.scala`

<img width="1594" alt="image" src="https://user-images.githubusercontent.com/18601388/188572899-5d41a01e-525f-419d-a575-5b9167b5aeb3.png">
<img width="1024" alt="image" src="https://user-images.githubusercontent.com/18601388/188573083-4098e275-c07f-41cf-b0a5-569a3795cfe3.png">

Now, this is curious. If I add another source in the same package, one level higher, BSP will set the sources dir one level higher as well after a reload.

```
▶ tree -A
.
└── src
    └── com
        └── github
            ├── AnotherObject.scala
            └── gedochao
                └── Main.scala

4 directories, 2 files
```

<img width="1192" alt="image" src="https://user-images.githubusercontent.com/18601388/188574329-6e8843dc-96ba-48e1-ad94-6059e31663ca.png">
<img width="1025" alt="image" src="https://user-images.githubusercontent.com/18601388/188574397-99710fa4-885b-4b97-8ca1-1f528328285d.png">

So I'm assuming it's setting the sources directory to the parent directory of the source closest to root.

**Known workarounds**
I can theoretically reset the sources directory to the package root's parent directory, but I'd have to redo it after every single BSP reload, which is simply impractical.
I can also add a mock source file outside of the package (in `(default)` package), so that the source root is picked where it should, but that's not ideal, either.
So you basically end up ignoring it and accepting that the project structure identified by IntelliJ isn't correct.

**Expected behaviour**
BSP should set  the sources directory in IntelliJ to the parent of the package root.

<img width="1181" alt="image" src="https://user-images.githubusercontent.com/18601388/188575807-add2df3e-31e5-4ff9-bcd3-9b1f823f729b.png">
<img width="1023" alt="image" src="https://user-images.githubusercontent.com/18601388/188575927-6c8c48d0-51a0-4606-93d6-d0a1fa6ae2d3.png">

**Comment by unkarjedy** [comment]
@Gedochao 
AFAIU scala-cli doesn't have a notion of `src` folder (If it does, please point me to the docs)
Seems like scala-cli bsp server should always report the "proejct" root as a source root.
And in your example, `src` should be treated as a part of a package.
![image](https://user-images.githubusercontent.com/3989292/220370312-50aa8114-e8d2-4a6c-8134-78ccf642ad86.png)

**Comment by Gedochao** [comment]
@unkarjedy
Sorry for the late response.
Long-post incoming.
> AFAIU scala-cli doesn't have a notion of src folder (If it does, please point me to the docs)

I am assuming you mean that Scala CLI does not have configurable `sources` roots for particular kinds of sources, like i.e. SBT does (corresponding SBT doc: https://www.scala-sbt.org/1.x/docs/Howto-Customizing-Paths.html)
Just to illustrate we're talking about the same thing, I'll give a corresponding example in SBT.

-----------

What I mean is, for SBT I could configure a project similar to the example above and import it via BSP by adding the following line to `build.sbt`
```scala
Compile / scalaSource := baseDirectory.value / "src"
```
and then I have a working project importable via BSP, no problem.

![image](https://user-images.githubusercontent.com/18601388/221596058-4c687576-6354-443d-ab16-85af67926462.png)

This is indeed because it's the SBT user's responsibility to override the correct folder.
Now let's take an extra look at SBT's BSP response for `buildTarget/sources`
```
[Trace - 02:25:30 PM] Received response 'buildTarget/sources - (4)' in 38ms
Result: {
  "items": [
    {
      "target": {
        "uri": "file:/Users/pchabelski/IdeaProjects/example-sbt/#root/Compile"
      },
      "sources": [
        {
          "uri": "file:/Users/pchabelski/IdeaProjects/example-sbt/src/",
          "kind": 2,
          "generated": false
        },
        {
          "uri": "file:/Users/pchabelski/IdeaProjects/example-sbt/src-3",
          "kind": 2,
          "generated": false
        },
        {
          "uri": "file:/Users/pchabelski/IdeaProjects/example-sbt/src/main/java",
          "kind": 2,
          "generated": false
        },
        {
          "uri": "file:/Users/pchabelski/IdeaProjects/example-sbt/target/scala-3.2.2/src_managed/main",
          "kind": 2,
          "generated": true
        }
      ]
    },
   ( ... )
  ]
}
Error: null
```
I know, long example, but it illustrates that SBT passes all of its configured sources directories (whether they're empty or not).

--------

Now, Scala CLI does not have a configurable sources directory (as you noted), and neither does it operate on directories - instead, it operates on individual source files.
Looking at Scala CLI's BSP `buildTarget/sources` response for the corresponding example project:

```
[Trace - 02:30:38 PM] Received response 'buildTarget/sources - (4)' in 17ms
Result: {
  "items": [
    {
      "target": {
        "uri": "file:/Users/pchabelski/IdeaProjects/scala-cli-tests/bsp-with-packages/.scala-build/?id\u003dproject_bd2c96d2de"
      },
      "sources": [
        {
          "uri": "file:///Users/pchabelski/IdeaProjects/scala-cli-tests/bsp-with-packages/src/com/github/gedochao/Main.scala",
          "kind": 1,
          "generated": false
        }
      ]
    },
    {
      "target": {
        "uri": "file:/Users/pchabelski/IdeaProjects/scala-cli-tests/bsp-with-packages/.scala-build/?id\u003dproject_bd2c96d2de-test"
      },
      "sources": []
    }
  ]
}
Error: null
```

As you can see, the individual source files are reported (and this is based on Bloop's approach to this - we actually base the response to this request on Bloop's).

------------

> Seems like scala-cli bsp server should always report the "project" root as a source root.
And in your example, src should be treated as a part of a package.

Now this is where things get complicated. We can't really arbitrarily report the project root as the source root in BSP, because the project root is not the source root - we do not have a source root, at all, we operate on single files. 
And those single files can be spread between directories, or even disk partitions, if the user so chooses.

Now, the project root - we do have a [dedicated doc for how we set where the project root is established](https://scala-cli.virtuslab.org/docs/reference/root-dir) (and thus, the `.bsp` directory).

Why not establish the common roots for all the on-disk inputs in the project, then? Well, The common root might include source files which are not, in fact, part of the project. Thus, we would report entire directories as `buildTarget/sources`, while in fact we build only a subset of its contents. Not only we'd report something false, we'd be confusing anyone looking at the logs.

To illustrate this with an example:

```sh
▶ tree
.
├── project-root
│   └── project.scala
├── src1
│   └── com
│       └── github
│           └── gedochao
│              └── Message.scala
├── src2
│   └── com
│       └── github
│           └── gedochao
│               └── Main.scala
└── src3
    └── StuffNotNeededOnTheClasspath.scala

11 directories, 6 files

```
Initializing the project:

```sh
scala-cli setup-ide src1 src2 project-root
# Wrote configuration file for ide in: /Users/pchabelski/IdeaProjects/scala-cli-tests/bsp-with-packages-complex-project-root/project-root/.bsp/scala-cli.json
```

Do note that even though `project-root` is passed last, it's established as the project root because it contains the `project.scala` (optional) configuration file.

Now when importing this into IntelliJ, I get:

<img width="1726" alt="image" src="https://user-images.githubusercontent.com/18601388/221791970-3b10e2b4-7629-4781-9207-a3d1358a98d6.png">

Which is not bad at all, since IntelliJ actually understands the disconnected locations of sources in the project and shows them. Code completion, all the major stuff is read correctly. Except, of course, for the weird package warning, along with the IDE arbitrarily treating the parent directory of each individual source as the source root.

<img width="798" alt="image" src="https://user-images.githubusercontent.com/18601388/221792453-86dd2164-96e4-448b-b1c0-a3b1d2d6ec82.png">

<img width="826" alt="image" src="https://user-images.githubusercontent.com/18601388/221792561-583aed27-0fe6-473f-a9db-9f8b468d9615.png">

TL;DR
Now, going back to @unkarjedy's suggestion, it's pretty obvious we can't just use the project root. In the example, it only contains the configuration file. Not useful.

We could establish the common root for all the sources, but that would catch everything in between. In the example, that'd be `.`, which would also catch `src3/StuffNotNeededOnTheClasspath.scala`. Not great. Not to mention establishing a common root may not even be possible, since the sources may be on different disk partitions.

We could special case IntelliJ, establish local roots where able to, without catching anything extra, and report that. That will not always be possible (i.e. in projects where some individual files of a directory are included in the build, but not all). And then, the packages would still remain off, since we'd still `guess` where the local source root should be. So perhaps not even worth the effort, since the bug would remain, just transform a bit.

I'd ask if it's possible for IntelliJ to understand individual files rather than directories, but I know it isn't.

In those circumstances, what I think would be best is if it was IntelliJ (well, the scala plugin) which would guess the source root correctly when a direct file path is given via BSP, based on the declared package and available file path. But that'd require a contribution to the IntelliJ Scala plugin, which I'm not too familiar with. So not sure how feasible that is.

Finally, we could leave things as they are, since everything technically works, except for the wonky structures in the project directory tree view and the false positive warnings about paths not matching packages. Code completion technically works code runs, despite the mixed signals sent to the user...
But yeah, we'd rather improve the UX in some way, if at all possible.

@unkarjedy what direction would you suggest, in this context?

**Comment by julienrf** [comment]
> I'd ask if it's possible for IntelliJ to understand individual files rather than directories, but I know it isn't.

This makes sense to me since it is very common to have Scala scripts as standalone files.

**Comment by unkarjedy** [comment]
For a start, we agreed to disable "Package name doesn't correspond to the file path" inspection 
for scala-cli projects imported via BSP

>which would guess the source root correctly when a direct file path is given via BSP, **based on the declared package and available file path**

I think such behavior could confuse even more taking that in scala-cli package name / directory structure correspondence doesn't make any sense.

>I'd ask if it's possible for IntelliJ to understand individual files rather than directories, but I know it isn't.

In theory, it could be possible, but it requires some experimenting.
We will try to experiment during 2023.2 release.

**Comment by Gedochao** [comment]
Thanks for the help, @unkarjedy!

> For a start, we agreed to disable "Package name doesn't correspond to the file path" inspection
for scala-cli projects imported via BSP

Awesome, this should leave Scala CLI users slightly less confused when working in IntelliJ, so it'll be a start.

>> which would guess the source root correctly when a direct file path is given via BSP, based on the declared package and available file path
>
> I think such behavior could confuse even more taking that in scala-cli package name / directory structure correspondence doesn't make any sense.

I suppose that's the cost of the project structure not being enforced by the language itself... Since the directory structure and package are only loosely tied in Scala, we prefer to keep it flexible in Scala CLI...

> > I'd ask if it's possible for IntelliJ to understand individual files rather than directories, but I know it isn't.
> 
> In theory, it could be possible, but it requires some experimenting. 
> We will try to experiment during 2023.2 release.

This makes me happy, looking forward to the results. 🚀 

I'm leaving this issue open, please go ahead and comment and link any relevant developments, so that we can track it all in one spot.

**Comment by unkarjedy** [comment]
>I'm leaving this issue open, please go ahead and comment and link any relevant developments, so that we can track it all in one spot

I think we can close this one in favor of https://youtrack.jetbrains.com/issue/SCL-19718 (and https://youtrack.jetbrains.com/issue/SCL-21048/scala-cli-disable-Package-name-doesnt-correspond-to-the-file-path-inspection-for-scala-files)
It's Scala Plugin who processes reported source files and calculates a single source folder.
BTW the logic is located here:
https://github.com/JetBrains/intellij-scala/blob/6ac4930da54499bb89e185a07517a1f934d5ebb8/bsp/src/org/jetbrains/bsp/project/importing/BspResolverLogic.scala#L345

**Comment by Gedochao** [comment]
okay then, let's track it under [SCL-19718](https://youtrack.jetbrains.com/issue/SCL-19718), I'm closing this for now.
